src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/RelationshipProcessor.cs

```diff
@@ -0,0 +1,247 @@
+ * caused by bugs or programming or logic errors or otherwise);
+ *
+ * A. any offer, purchase, holding, use, sale, exchange or transmission of any
+ * cryptographic keys, tokens or assets created, exchanged, stored or arising from any
+ * interaction with the Work;
+ *
+ * B. any failure in a transmission or loss of any token or assets keys or other digital
+ * artefacts due to errors in transmission;
+ *
+ * C. bugs, hacks, logic errors or faults in the Work or any communication;
+ *
+ * D. system software or apparatus including but not limited to losses caused by errors
+ * in holding or transmitting tokens by any third-party;
+ *
+ * E. breaches or failure of security including hacker attacks, loss or disclosure of
+ * password, loss of private key, unauthorised use or misuse of such passwords or keys;
+ *
+ * F. any losses including loss of anticipated savings or other benefits resulting from
+ * use of the Work or any changes to the Work (however implemented).
+ *
+ * You are solely responsible for; testing, validating and evaluation of all operation
+ * logic, functionality, security and appropriateness of using the Work for any commercial
+ * or non-commercial purpose and for any reproduction or redistribution by You of the
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+using RadixDlt.NetworkGateway.Abstractions;
+using RadixDlt.NetworkGateway.Abstractions.Model;
+using RadixDlt.NetworkGateway.PostgresIntegration.Models;
+using System;
+using System.Diagnostics;
+using System.Linq;
+using CoreModel = RadixDlt.CoreApiSdk.Model;
+
+namespace RadixDlt.NetworkGateway.PostgresIntegration.LedgerExtension;
+
+internal class RelationshipProcessor
+{
+    private readonly ReferencedEntityDictionary _referencedEntities;
+
+    public RelationshipProcessor(ReferencedEntityDictionary referencedEntities)
+    {
+        _referencedEntities = referencedEntities;
+    }
+
+    public void ScanUpsert(CoreModel.Substate substateData, ReferencedEntity referencedEntity, long stateVersion)
+    {
+        if (substateData is CoreModel.IRoyaltyVaultHolder royaltyVaultHolder && royaltyVaultHolder.TryGetRoyaltyVault(out var rv))
+        {
+            _referencedEntities
+                .Get((EntityAddress)rv.EntityAddress)
+                .PostResolveConfigure((InternalFungibleVaultEntity e) =>
+                {
+                    e.AddCorrelation(EntityRelationship.Royalty, referencedEntity.DatabaseId);
+                });
+        }
+
+        if (substateData is CoreModel.TypeInfoModuleFieldTypeInfoSubstate typeInfoSubstate)
+        {
+            switch (typeInfoSubstate.Value.Details)
+            {
+                case CoreModel.ObjectTypeInfoDetails objectDetails:
+                    referencedEntity.PostResolveConfigure((ComponentEntity e) =>
+                    {
+                        e.AssignedModuleIds = objectDetails.ModuleVersions
```

Comment:
it's not related to relationships. maybe we have to process same substate in multiple places to keep it clean.

Reply:
Ah, I noticed that myself and wanted to change that before pushing... obviously I must have forgotten about about that before I pushed.